src/commonMain/kotlin/teksturepako/pakku/cli/cmd/Diff.kt

```diff
@@ -15,43 +17,88 @@ class Diff : CliktCommand("Diff projects in modpack")
-                {
-                    newProject
-                }
-                else null
-            }
-            .mapNotNull { it.name.values.firstOrNull() }
-
+        val added = allNewProjects.mapNotNull { newProject ->
+            if (allOldProjects containsNotProject newProject)
+            {
+                newProject
+            } else null
+        }.mapNotNull { it.name.values.firstOrNull() }
         added.forEach { terminal.success("+ $it") }
 
-        val removed = oldLockFile.getAllProjects()
-            .mapNotNull { oldProject ->
-                if (newLockFile.getAllProjects() containsNotProject oldProject)
+        val removed = allOldProjects.mapNotNull { oldProject ->
+            if (allNewProjects containsNotProject oldProject)
+            {
+                oldProject
+            } else null
+        }.mapNotNull { it.name.values.firstOrNull() }
+        removed.forEach { terminal.danger("- $it") }
+
+        val updated: MutableList<String> = mutableListOf()
+        val oldFiles: MutableList<String> = mutableListOf()
+        val newFiles: MutableList<String> = mutableListOf()
+        for (oldProject in allOldProjects)
+        {
+            /** We only care about projects, which previously also existed **/
+            if (allNewProjects containsProject oldProject)
+            {
+                for (newProject in allNewProjects)
                 {
-                    oldProject
+                    /** Everything inside if-block is the same project **/
+                    if (newProject isAlmostTheSameAs oldProject)
+                    {
+                        val oldProjectFiles = oldProject.files
+                        val newProjectFiles = newProject.files
+                        /** If the project files are not identical, means that files have changed **/
+                        if (oldProjectFiles != newProjectFiles)
+                        {
+                            /** On multiloader modpacks, the mod might have been added for another loader, causing the
+                             * files key to have changed without an actual update
+                             **/
+                            if (oldProjectFiles.firstOrNull()?.hashes?.get("sha1") != newProjectFiles.firstOrNull()?.hashes?.get("sha1")                            )
+                            {
+                                if (detailedUpdate)
+                                {
+                                    oldFiles.add("${oldProjectFiles.firstOrNull()?.fileName}")
+                                    newFiles.add("${newProjectFiles.firstOrNull()?.fileName}")
+                                } else
+                                {
+                                    updated.add("${oldProject.name.values.firstOrNull()}")
+                                }
+                            }
+                        }
+                    }
```

Comment:
This is too much nested. You could break it into functions and continue early instead of nesting ifs...

Like this:
```kotlin
        for (oldProject in allOldProjects)
        {
            /** We only care about projects, which previously also existed **/
            if (allNewProjects containsNotProject oldProject) continue
            
            x@ for (newProject in allNewProjects)
            {
                /** Everything inside if-block is the same project **/
                if (!newProject.isAlmostTheSameAs(oldProject)) continue@x

// and so on...
```
The `x@` is a label, so you can use `continue@x` inside the nested for loop.

If you want to do it even better, you could use `map` and `flatMap` to process the data.

Reply:
> /** Everything inside if-block is the same project **/

Also, some comments would probably stop making sense then.